src/integration/java/io/pinecone/integration/dataplane/UpsertAndDeleteTest.java

```diff
@@ -0,0 +1,240 @@
+package io.pinecone.integration.dataplane;
+
+import com.google.protobuf.Struct;
+import com.google.protobuf.Value;
+import io.pinecone.PineconeConnection;
+import io.pinecone.helpers.IndexManager;
+import io.pinecone.helpers.RandomStringBuilder;
+import io.pinecone.proto.*;
+import org.junit.jupiter.api.BeforeAll;
+import org.junit.jupiter.api.Test;
+
+import java.io.IOException;
+import java.util.Arrays;
+import java.util.List;
+import java.util.concurrent.ExecutionException;
+
+import static io.pinecone.helpers.BuildUpsertRequest.buildOptionalUpsertRequest;
+import static io.pinecone.helpers.BuildUpsertRequest.buildRequiredUpsertRequest;
+import static org.junit.jupiter.api.Assertions.assertEquals;
+import static org.junit.jupiter.api.Assertions.assertThrows;
+
+public class UpsertAndDeleteTest {
+    private static VectorServiceGrpc.VectorServiceBlockingStub blockingStub;
+    private static VectorServiceGrpc.VectorServiceFutureStub futureStub;
+    private static final int dimension = 3;
+
+    @BeforeAll
+    public static void setUp() throws IOException, InterruptedException {
+        PineconeConnection connection = new IndexManager().createIndexIfNotExists(dimension);
+        blockingStub = connection.getBlockingStub();
+        futureStub = connection.getFutureStub();
+    }
+
+    @Test
+    public void UpsertVectorsAndDeleteByIdSyncTest() throws InterruptedException {
+        // Upsert vectors with required parameters
+        String namespace = RandomStringBuilder.build("ns", 8);
+        List<String> upsertIds = Arrays.asList("v1", "v2", "v3");
+        UpsertResponse upsertResponse = blockingStub.upsert(buildRequiredUpsertRequest(upsertIds, namespace));
+
+        // Get vector count before deleting vectors
+        DescribeIndexStatsRequest describeIndexRequest = DescribeIndexStatsRequest.newBuilder().build();
+        DescribeIndexStatsResponse describeIndexStatsResponse = blockingStub.describeIndexStats(describeIndexRequest);
+        int startVectorCount = describeIndexStatsResponse.getNamespacesMap().get(namespace).getVectorCount();
+        assertEquals(startVectorCount, upsertResponse.getUpsertedCount());
+
+        // Delete 1 vector
+        String[] idsToDelete = new String[]{upsertIds.get(0)};
+        DeleteRequest deleteRequest = DeleteRequest.newBuilder()
+                .setNamespace(namespace)
+                .addAllIds(Arrays.asList(idsToDelete))
+                .setDeleteAll(false)
+                .build();
+        blockingStub.delete(deleteRequest);
+        Thread.sleep(3500);
```

Comment:
Deletion is not atomic, so I don't want to take a chance where the deletion was not completed and the verification using describeIndexStats starts. Dont remember if I saw it failed atleast once but this can happen when integration tests are ran on every commit. This is where our internal conversation on making integration test bullet proof is related (instead of a thread.sleep, retry with exponential back-off).